doc/developer/design/20210728_error_handling_surfacing_connector_failures.md

```diff
@@ -0,0 +1,189 @@
+- introduce connector lifecycle and surface it to users through system tables
+
+## Non-Goals
+
+- introduce more structured error messages, see
+  https://github.com/MaterializeInc/materialize/discussions/5340
+
+## Description
+
+The actual proposal is [way down](#proposed-changes) but it helps to first
+understand the current state.
+
+### Current State
+
+#### Where/when do we interact with external systems?
+
+- When creating a source/sink:
+  - (coord) purification
+- When creating a source/sink or when materialized is restarted:
+  - (coord) initial connector setup, for example:
+  - (coord) reading Kafka consistency topic
+  - (coord) publishing schemas
+  - (coord) creating topics
+  - (coord) creating output file, checking input file
+- Continuously:
+  - (coord) metadata loops, for example fetching new Kafka partitions, listening
+   on BYO topics, listening on SQS for S3 updates
+  - (dataflow) actual data ingest and writing
+
+#### Failures during purification
+
+In general, failures during purification lead to the statement being aborted,
+nothing gets added to the catalog. This seems fair, because we cannot store
+un-purified statements in the catalog and purification happens infrequently,
+only when a statement is issued.
+
+Most purification logic doesn't specify timeouts, only S3 and Kinesis have a 1
+second timeout for the validation call. I believe none of the purification
+checks do retries
+
+#### Failures During Source/Sink Creation
+
+The behaviour is very similar (the same?) to failures during purification:
+nothing gets added to the catalog when there are failures.
+
+Some calls have custom timeouts, some don't. I believe none of the calls have
+retries.
+
+#### Failures During "Normal Operations"
+
+Failures during execution don't bring down materialized. If failures happen,
+they are only surfaced in the log. Materialized doesn't "know" about the state
+of a connector.
+
+There are no global config options for timeouts/retries/back-off interval.
+
+#### Failures During Restart (restarting a materialized with things in the catalog)
+
+Failures during sink (re-)creation kill materialized when restarting with a
+filled catalog.  Failures during source creation are fine.
+
+The reason is that (re-)creating a source runs no source-specific code in the
+coordinator. Everything happens in dataflow, where we have retries. This is not
+true for sinks.
+
+### Proposed Changes
+
+The reaction to failures at various stages should uniform across different
+connectors and informed by these rules:
+
+ - Failures during connector creation (which includes purification) immediately
+   report an error back to the user. No catalog entry is created.
+ - Failures during runtime need to be surfaced to the user in system tables.
+ - Failures that occur during a restart with an already filled catalog must not
+   bring down Materialize. Instead, errors that occur must be reported to the
+   user for each individual connector.
+
```

Comment:
failures in the source should also cause selecting from the source to start erroring out, rather than continuing to return stale or no data. This should be in addition to sufracing that same error in a `mz_` table.

Reply:
Do you mean fatal errors, like a malformed message that we can't parse. Or transient errors, like a Kafka broker not being reachable. For the former, we should already prevent querying, but you're right that we don't do anything about the latter right now.